### ISSUE TITLE
Consider removing the OrderedSerialization on Map/Set

### ISSUE THREAD

**Issue by johnynek** [twitter/scalding#1331]
SortedMap/SortedSet make sense, and users might not expect that just serializing requires sorting, which it does.

We could make a `Serialization` for Map/Set that does not sort, which is what you want to use when using Map/Set in a value. It really does not make sense to use Map/Set as a key, you should use SortedMap/SortedSet (and even then for SortedMap only when the value also has an OrderedSerialization).

**Comment by isnotinvain** [comment]
+1, as is users will get garbage results if they use a map or set in a key right?

**Comment by ianoc** [comment]
No, the results now will be fine. It will sort them as if they had done SortedMap or SortedSet

**Comment by isnotinvain** [comment]
oh -- then... isn't this a good thing?

**Comment by johnynek** [comment]
If you want to sort a Map, why not use a `SortedMap`? The problem is that even when you serialize a `Map` but don't want to compare it, you are going to sort it, which is not free (a `log N` penalty could be pretty big).

I feel like this is a bit clever rather than explicit. It works, but it makes significant performance costs invisible.

**Comment by isnotinvain** [comment]
oh is this to support a non-ordered serialization for Maps / Sets when they're used as values? Makes sense to me

**Comment by johnynek** [comment]
Right. We have a `Serialization[T]` typeclass, and they could be fast for
Map and Set and keep the sorted approach for SortedMap/SortedSet.

On Thu, Jun 18, 2015 at 2:38 PM, Alex Levenson <notifications@github.com>
wrote:

> oh is this to support a non-ordered serialization for Maps / Sets when
> they're used as values? Makes sense to me
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/twitter/scalding/issues/1331#issuecomment-113329465>.
>



-- 
Oscar Boykin :: @posco :: http://twitter.com/posco

**Comment by ianoc** [comment]
Right now we only do macros on one type, OrderedSerialization which supplies Orderings and Serializations . Today we only use our trait for key's in group by's, or secondary sorts now. 

I'm just not sure its worth removing, there is a pretty reasonable logic that if you want to put it in the key then having to maintain a sorted map on every update is much more expensive than just doing it at the serialization point when exiting the jvm. 

Is anyone using this to get a serialization of a map and having an issue or is this mostly hypothetical for now?

**Comment by johnynek** [comment]
Well this is for the case where we imagine requiring Serialization on the
values, and OrderedSerialization on the keys, and never use Kryo by default
in the Typed-API.

If we did this now, .toSet would get a lot slower, for instance, because we
would sort the values.

On Thu, Jun 18, 2015 at 2:42 PM, ianoc <notifications@github.com> wrote:

> Right now we only do macros on one type, OrderedSerialization which
> supplies Orderings and Serializations . Today we only use our trait for
> key's in group by's, or secondary sorts now.
>
> I'm just not sure its worth removing, there is a pretty reasonable logic
> that if you want to put it in the key then having to maintain a sorted map
> on every update is much more expensive than just doing it at the
> serialization point when exiting the jvm.
>
> Is anyone using this to get a serialization of a map and having an issue
> or is this mostly hypothetical for now?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/twitter/scalding/issues/1331#issuecomment-113329778>.
>



-- 
Oscar Boykin :: @posco :: http://twitter.com/posco

**Comment by ianoc** [comment]
Would it though? I think it'd be interesting to try see that, once you are at a touch point of serializing then the cost of sorting whatever is in a single record is unlikely that large is it?  If we change these we loose our 'all common case classes just work' setup we are at now, along with scrooge structures won't have orderings anymore where they contain maps or sets (maps anyway seem pretty common).
